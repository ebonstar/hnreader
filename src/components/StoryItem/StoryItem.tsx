import timeago from "epoch-timeago";

export type Story = {
  id: number;
  title: string;
  url?: string;
  score: number;
  descendants: number;
  time: number;
};

const backgroundColours = [
  "bg-yellow-300",
  "bg-yellow-300",
  "bg-lime-300",
  "bg-lime-300",
  "bg-teal-300",
  "bg-cyan-300",
  "bg-cyan-300",
  "bg-stone-200",
  "bg-rose-400",
  "bg-emerald-300",
];

const topPositions = ["top-1", "top-3"];

const scales = ["", "scale-110", "scale-125"];

const extraInfoClasses =
  "absolute hidden group-hover:block bg-black text-white";

export function StoryItem({ story }: { story: Story }) {
  const hnLink = `https://news.ycombinator.com/item?id=${story.id}`;
  const background = backgroundColours[story.id % 10];
  const top = topPositions[story.score % 2];
  const scale =
    story.score > 300 ? scales[2] : story.score > 150 ? scales[1] : scales[0];
  return (
    <div
      className={`group relative mb-6 px-8 py-4
          ${background} border-4 border-black ${scale}
          hover:rotate-3 hover:z-10 transition-all`}
      key={story.id}
    >
      <div className={`${extraInfoClasses} ${top} -left-12 p-2 font-bold`}>
        â‡§ {story.score}
      </div>
      <a href={story.url ?? hnLink} target="_blank">
        <div className="text-xl font-bold">{story.title}</div>
        <div className="text-m">{timeago(story.time * 1000)}</div>
      </a>
      <a
        className={`${extraInfoClasses} -bottom-6 -right-2 px-4 py-1 hover:underline`}
        href={hnLink}
        target="_blank"
      >
        {story.descendants ?? 0}{" "}
        {story.descendants === 1 ? " comment" : "comments"}
      </a>
    </div>
  );
}
