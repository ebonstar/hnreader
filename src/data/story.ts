import { Story } from "../components/StoryItem";
import { fetchUrl } from "./api";

export const fetchTopStoryIds = async () => {
  try {
    return fetchUrl<number[]>("topstories.json");
  } catch (err) {
    console.error("Could not load top stories", err);
    throw err;
  }
};

export const fetchStory = async (id: number) => {
  try {
    return fetchUrl<Story>(`item/${id}.json`);
  } catch (err) {
    console.error("Could not load story", id, err);
    throw err;
  }
};
