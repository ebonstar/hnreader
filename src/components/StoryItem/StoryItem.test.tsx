import { render, screen } from "@testing-library/react";
import { HN_ITEM_LINK, StoryItem } from "./StoryItem";
import story from "mocks/item.json";

test("storyItem matches snapshot", () => {
  const storyItem = render(<StoryItem story={story} />);
  expect(storyItem).toMatchSnapshot();
});

test("story with no external url links to hacker news page", () => {
  const { url, ...storyWithoutUrl } = story;
  render(<StoryItem story={storyWithoutUrl} />);

  const storyLink: HTMLAnchorElement = screen.getByRole("link", {
    name: /story title/,
  });

  expect(storyLink.href).toContain(HN_ITEM_LINK);
});
