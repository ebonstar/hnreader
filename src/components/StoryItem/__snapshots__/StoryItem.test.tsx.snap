// Vitest Snapshot v1

exports[`storyItem matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="group relative mb-6 px-8 py-4
        bg-lime-300 border-4 border-black scale-110 rotate-3
        hover:rotate-0 hover:z-10 transition-all"
      >
        <div
          class="absolute hidden group-hover:block bg-black text-white top-12 -left-12 p-2 font-bold"
        >
          ⇧ 
          200
        </div>
        <a
          href="http://google.com/q?=0"
          target="_blank"
        >
          <div
            class="text-xl font-bold"
          >
            This is a story title
          </div>
          <div
            class="text-m"
          >
            3 days ago
          </div>
        </a>
        <a
          class="absolute hidden group-hover:block bg-black text-white -bottom-6 -right-2 px-4 py-1 hover:underline"
          href="https://news.ycombinator.com/item?id=0"
          target="_blank"
        >
          20
           
          comments
        </a>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="group relative mb-6 px-8 py-4
        bg-lime-300 border-4 border-black scale-110 rotate-3
        hover:rotate-0 hover:z-10 transition-all"
    >
      <div
        class="absolute hidden group-hover:block bg-black text-white top-12 -left-12 p-2 font-bold"
      >
        ⇧ 
        200
      </div>
      <a
        href="http://google.com/q?=0"
        target="_blank"
      >
        <div
          class="text-xl font-bold"
        >
          This is a story title
        </div>
        <div
          class="text-m"
        >
          3 days ago
        </div>
      </a>
      <a
        class="absolute hidden group-hover:block bg-black text-white -bottom-6 -right-2 px-4 py-1 hover:underline"
        href="https://news.ycombinator.com/item?id=0"
        target="_blank"
      >
        20
         
        comments
      </a>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
